File: packages/node/src/evidence.ts
import { createPublicKey, verify as cryptoVerify } from 'node:crypto'
import type { EvidenceKey, EvidenceSignature } from './types'

/**
 * Verificação OFFLINE da evidência assinada, sem depender da Catalisa
 * (`src/biometrics/services/evidence-signature.service.ts`):
 *
 *   mensagem   = `${sessionId}|${attempt}|${bundleHash}|${signedAt}`
 *   assinatura = Ed25519 (RFC 8032) com a chave da organização, em base64
 *   chave      = `GET /evidence-keys` → `publicKeyPem` do `keyId` da assinatura (chaves aposentadas continuam publicadas)
 *
 * `attempt` é o `attempt` do envelope da sessão (`GET /sessions/:id`).
 */
export function evidenceMessage(sessionId: string, attempt: number, bundleHash: string, signedAt: string): string {
  return `${sessionId}|${attempt}|${bundleHash}|${signedAt}`
}

export interface EvidenceToVerify {
  sessionId: string
  attempt: number
  bundleHash: string
  signature: EvidenceSignature
}

export interface OfflineVerification {
  valid: boolean
  keyId: string
  /** `true` se não há chave publicada com esse `keyId`. */
  unknownKey: boolean
  retiredAt: string | null
}

export function verifyEvidenceSignature(ev: EvidenceToVerify, keys: ReadonlyArray<EvidenceKey> | Record<string, string>): OfflineVerification {
  const keyId = ev.signature.keyId
  const list: EvidenceKey[] = Array.isArray(keys)
    ? (keys as EvidenceKey[])
    : Object.entries(keys as Record<string, string>).map(([k, pem]) => ({ keyId: k, publicKeyPem: pem, createdAt: '', retiredAt: null }))
  const key = list.find((k) => k.keyId === keyId)
  if (!key) return { valid: false, keyId, unknownKey: true, retiredAt: null }
  if (ev.signature.alg !== 'Ed25519') return { valid: false, keyId, unknownKey: false, retiredAt: key.retiredAt }
  let valid = false
  try {
    valid = cryptoVerify(
      null,
      Buffer.from(evidenceMessage(ev.sessionId, ev.attempt, ev.bundleHash, ev.signature.signedAt), 'utf8'),
      createPublicKey(key.publicKeyPem),
      Buffer.from(ev.signature.value, 'base64')
    )
  } catch {
    valid = false
  }
  return { valid, keyId, unknownKey: false, retiredAt: key.retiredAt }
}

File: packages/python/src/catalisa_biometrics/evidence.py
"""OFFLINE verification of the signed evidence (Ed25519), without depending on Catalisa.

    message = f"{sessionId}|{attempt}|{bundleHash}|{signedAt}"
    key     = GET /evidence-keys → publicKeyPem of the signature's keyId
"""

from __future__ import annotations

import base64
import binascii
from typing import Any, Dict, Iterable, Mapping, Union

from . import _ed25519


def evidence_message(session_id: str, attempt: int, bundle_hash: str, signed_at: str) -> str:
    return f"{session_id}|{attempt}|{bundle_hash}|{signed_at}"


def verify_evidence_signature(
    session_id: str,
    attempt: int,
    bundle_hash: str,
    signature: Mapping[str, Any],
    keys: Union[Iterable[Mapping[str, Any]], Mapping[str, str]],
) -> Dict[str, Any]:
    key_id = signature.get("keyId")
    if isinstance(keys, Mapping):
        found = {"keyId": key_id, "publicKeyPem": keys.get(key_id), "retiredAt": None} if key_id in keys else None
    else:
        found = next((k for k in keys if k.get("keyId") == key_id), None)
    if not found or not found.get("publicKeyPem"):
        return {"valid": False, "keyId": key_id, "unknownKey": True, "retiredAt": None}
    valid = False
    if signature.get("alg") == "Ed25519":
        try:
            pub = _ed25519.load_public_key(found["publicKeyPem"])
            sig = base64.b64decode(signature.get("value", ""), validate=True)
            msg = evidence_message(session_id, attempt, bundle_hash, signature.get("signedAt", "")).encode("utf-8")
            valid = _ed25519.verify(pub, msg, sig)
        except (ValueError, binascii.Error):
            valid = False
    return {"valid": valid, "keyId": key_id, "unknownKey": False, "retiredAt": found.get("retiredAt")}
